### PR TITLE
[Form] DateTimeImmutable norm data in DateTime form types

### PR DESCRIPTION
**Pull request by vudaltsov**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

This PR suggests a way to use `\DateTimeImmutable` objects for norm data in `DateTimeType`, `DateType` and `TimeType`.

A complete BC layer for transformers is provided. All protected calls are also taken into account.

If this PR gets merged, we can then easily add `input=datetimeimmutable` with no model transformers to support `\DateTimeImmutable` model data.

### CODE REVIEW

**Comment by vudaltsov** [comment]
ping @xabbuh , @stof, @HeahDude

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToArrayTransformer.php [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+/**
+ * Transforms between a normalized time and a localized time string/array.
+ *
+ * @author Bernhard Schussek <bschussek@gmail.com>
+ * @author Florian Eckerstorfer <florian@eckerstorfer.org>
+ */
+class DateTimeImmutableToArrayTransformer extends BaseDateTimeTransformer
+{
+    private $pad;
+
```
Comment (Simperfit) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToArrayTransformer.php]: no need for this empty line

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToLocalizedStringTransformer.php [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+/**
+ * Transforms between a normalized time and a localized time string.
+ *
+ * @author Bernhard Schussek <bschussek@gmail.com>
+ * @author Florian Eckerstorfer <florian@eckerstorfer.org>
+ */
+class DateTimeImmutableToLocalizedStringTransformer extends BaseDateTimeTransformer
+{
+    private $dateFormat;
+
```
Comment (Simperfit) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToLocalizedStringTransformer.php]: same

**Review by Simperfit — CHANGES_REQUESTED** [review]
You need to add tests for all the new classes you created.

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToArrayTransformer.php (hunk 1 above)
Comment (vudaltsov) [reply src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeImmutableToArrayTransformer.php]: thanx, done

**Comment by vudaltsov** [comment]
Added tests.

**Comment by stof** [comment]
I'm not sure about the BC impact of changing the type of the norm data. I suggest adding the new input type first, without changing the norm data.

**Comment by vudaltsov** [comment]
[Here](https://github.com/symfony/symfony/pull/19889#issuecomment-283394385) @HeahDude posted:

> If a new major version of Symfony bumps the requirement to PHP 5.5+, the normalized data could then be switched to the DateTimeImmutable class to simplify the logic (which is currently cloning the DateTime to avoid modifying the original one.

That's what I tried to do in my PR. My change is pretty straightforward: I copied and renamed transformers and their tests and removed `clone`.

If we really want to make norm data immutable than I think we should do it first. If we don't - than yes, just add `input=datetimimmutable` and a corresponding transformer.

But with what I suggest immutable dates will be supported in the first place. Only `DateTime`s will be converted. In this way immutable norm data helps to follow the trend of replacing `DateTime` with `DateTimeImmutable`.

**Comment by stof** [comment]
you need to add the new allowed value for `input` option. Otherwise, immutable datetime are **not** supported (they are used as norm data, but cannot be used as model data, which is the requested feature)

**Comment by vudaltsov** [comment]
Sure! I planned to do that in a separate PR. Because it's gonna have a couple of new tests.
In this PR I just provide a ground for that by changing the norm data.

If you wish, I can add commits right here.

**Comment by vudaltsov** [comment]
@Simperfit, could you please review my tests?

File: src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/DateTimeImmutableToTimestampTransformerTest.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        $output->setTimezone(new \DateTimeZone('Asia/Hong_Kong'));
+
+        $this->assertDateTimeImmutableEquals($output, $reverseTransformer->reverseTransform($input));
+    }
+
+    public function testReverseTransformExpectsValidTimestamp()
+    {
+        $reverseTransformer = new DateTimeImmutableToTimestampTransformer();
+
+        $this->{method_exists($this, $_ = 'expectException') ? $_ : 'setExpectedException'}('Symfony\Component\Form\Exception\TransformationFailedException');
```
Comment (Simperfit) [inline src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/DateTimeImmutableToTimestampTransformerTest.php]: maybe you should add a new var for this.

**Review by Simperfit — COMMENTED** [review]
The tests looks good to me,

**Comment by vudaltsov** [comment]
Closed in favor of https://github.com/symfony/symfony/pull/25273.

**Pull request closed without merging**